infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java

```diff
@@ -144,15 +149,22 @@ public void apply(
 
     List<PodData> podsData = getPodDatas(componentObjects);
 
-    estimateCommandsMachineName(workspaceConfig, k8sComponent, podsData);
+    Map<String, Map<String, String>> podToContainerAndMachineName = estimateMachineNames(podsData);
```

Comment:
I don't like this `Map<String, Map<String, String>>`, it's very hard to follow. Does it really have to be map of maps ? I can't find any usage of it's keys.

Reply:
It's because containers in different pods may have same names. So need to store them separately. I will try to find way to simplify, but not sure if i could.

Reply:
I can propose to go in another direction. You don't really know pods and containers during commands machine name estimation, so it's what I tried to implement (but not tested):
```java
public void apply ...
   List<PodData> podsData = getPodDatas(componentObjects);

    if (Boolean.TRUE.equals(k8sComponent.getMountSources())) {
      applyProjectsVolumes(podsData, componentObjects);
    }

    applyEntrypoints(k8sComponent.getEntrypoints(), componentObjects);

    Map<String, MachineConfigImpl> machineConfigs =
        k8sComponent.getAlias() != null
            ? prepareMachineConfigs(podsData, k8sComponent.getAlias())
            : emptyMap();

    linkCommandsToMachineName(workspaceConfig, k8sComponent, machineConfigs.keySet());
  ...

  /**
   * Creates map of machine names and corresponding {@link MachineConfigImpl} with component alias
   * attribute set.
   */
  private Map<String, MachineConfigImpl> prepareMachineConfigs(List<PodData> podsData,
      @Nullable String componentAlias) {
    Map<String, MachineConfigImpl> machineConfigs = new HashMap<>();
    for (PodData podData : podsData) {
      for (Container container : podData.getSpec().getContainers()) {
        String machineName = machineName(podData, container);

        MachineConfigImpl config = new MachineConfigImpl();
        if (Strings.isNullOrEmpty(componentAlias)) {
          config.getAttributes().put(DEVFILE_COMPONENT_ALIAS_ATTRIBUTE, componentAlias);
        }

        machineConfigs.put(machineName, config);
      }
    }
    return machineConfigs;
  }

/**
   * Set {@link Command#MACHINE_NAME_ATTRIBUTE} to commands which are configured in the specified
   * component.
   *
   * <p>Machine name will be set only if the specified recipe objects has the only one container.
   */
  private void linkCommandsToMachineName(
      WorkspaceConfig workspaceConfig,
      Component component,
      Set<String> machinesNames) {
    List<? extends Command> componentCommands =
        workspaceConfig
            .getCommands()
            .stream()
            .filter(
                c ->
                    component.getAlias() != null
                        && component
                            .getAlias()
                            .equals(
                                c.getAttributes().get(Constants.COMPONENT_ALIAS_COMMAND_ATTRIBUTE)))
            .collect(toList());
    if (componentCommands.isEmpty()) {
      return;
    }

    if (machinesNames.size() != 1) {
      // many or no pods - can't estimate the name because of ambiguity or lack of information
      return;
    }

    String machineName = machinesNames.iterator().next();
    componentCommands.forEach(c -> c.getAttributes().put(MACHINE_NAME_ATTRIBUTE, machineName));
  }
```

```

Reply:
Ok i will apply the approach with some minot edits, thanks